src/core/ext/transport/inproc/inproc_transport.c

```diff
@@ -0,0 +1,938 @@
+  s->read_closure_needed = false;
+  grpc_closure_init(&s->read_closure, read_state_machine, s,
+                    grpc_schedule_on_exec_ctx);
+  s->t = t;
+  s->closure_at_destroy = NULL;
+  s->num_linked_mds = 0;
+  s->space_linked_mds = 0;
+  s->linked_md_array = NULL; /* grow this when relevant */
+
+  s->initial_md_sent = s->trailing_md_sent = s->initial_md_recvd =
+      s->trailing_md_recvd = false;
+
+  s->cancel_error = GRPC_ERROR_NONE;
+  s->write_buffer_cancel_error = GRPC_ERROR_NONE;
+  s->deadline = gpr_inf_future(GPR_CLOCK_MONOTONIC);
+  s->write_buffer_deadline = gpr_inf_future(GPR_CLOCK_MONOTONIC);
+
+  if (!server_data) {
+    ref_transport(t);
+    inproc_transport *st = t->other_side;
+    ref_transport(st);
+    s->other_side = NULL;  // will get filled in soon
+    // Pass the client-side stream address to the server-side for a ref
+    ref_stream(s);  // ref it now on behalf of server side to avoid destruction
+    gpr_log(GPR_DEBUG, "calling accept stream cb %p %p", st->accept_stream_cb,
+            st->accept_stream_data);
+    (*st->accept_stream_cb)(exec_ctx, st->accept_stream_data, &st->base,
+                            (void *)s);
+  } else {
+    // This is the server-side and is being called through accept_stream_cb
+    inproc_stream *cs = (inproc_stream *)server_data;
+    s->other_side = cs;
+    // Ref the server-side stream on behalf of the client now
+    ref_stream(s);
+
+    // Now we are about to affect the other side, so lock the transport
+    gpr_mu_lock(&s->t->mu->mu);
+    cs->other_side = s;
+    // Now transfer from the other side's write_buffer if any to the to_read
+    // buffer
+    if (GRPC_SLICE_START_PTR(cs->write_buffer_initial_md) != NULL) {
+      grpc_slice_unref_internal(exec_ctx, s->to_read_initial_md);
+      s->to_read_initial_md = grpc_slice_ref(cs->write_buffer_initial_md);
+      grpc_slice_unref_internal(exec_ctx, cs->write_buffer_initial_md);
+      s->deadline = gpr_time_min(s->deadline, cs->write_buffer_deadline);
+    }
+    if (cs->write_buffer_message.length > 0) {
+      grpc_slice_buffer_swap(&cs->write_buffer_message, &s->to_read_message);
+      GPR_ASSERT(cs->write_buffer_message.count == 0);
+    }
+    if (GRPC_SLICE_START_PTR(cs->write_buffer_trailing_md) != NULL) {
+      grpc_slice_unref_internal(exec_ctx, s->to_read_trailing_md);
+      s->to_read_trailing_md = grpc_slice_ref(cs->write_buffer_trailing_md);
+      grpc_slice_unref_internal(exec_ctx, cs->write_buffer_trailing_md);
+    }
+    if (cs->write_buffer_cancel_error != GRPC_ERROR_NONE) {
+      s->cancel_error = GRPC_ERROR_REF(cs->write_buffer_cancel_error);
+    }
+
+    gpr_mu_unlock(&s->t->mu->mu);
+  }
+  return 0;  // return value is not important
+}
+
+static grpc_slice fill_in_metadata(const grpc_metadata_batch *metadata,
+                                   uint32_t flags) {
+  // Calculate the size needed to store the entire metadata block, including
```

Comment:
Any reason we're serializing at all here?

Why not just populate a metadata batch...

```c
static void send_metadata(const grpc_metadata_batch *src, grpc_metadata_batch *dst) {
  for (grpc_linked_mdelem *elem = src->list.head; elem != NULL;
        elem = elem->next) {
     grpc_metadata_batch_add_tail(..., elem->md);
   }
}
```